src/v/cluster/persisted_stm.cc

```diff
@@ -181,6 +182,96 @@ size_t file_backed_stm_snapshot::get_snapshot_size() const {
     return _snapshot_size;
 }
 
+kvstore_backed_stm_snapshot::kvstore_backed_stm_snapshot(
+  ss::sstring snapshot_name,
+  prefix_logger& log,
+  raft::consensus* c,
+  storage::kvstore& kvstore)
+  : _ntp(c->ntp())
+  , _name(snapshot_name)
+  , _snapshot_key(ssx::sformat("{}/{}", snapshot_name, c->ntp()))
+  , _log(log)
+  , _kvstore(kvstore) {}
+
+kvstore_backed_stm_snapshot::kvstore_backed_stm_snapshot(
+  ss::sstring snapshot_name,
+  prefix_logger& log,
+  model::ntp ntp,
+  storage::kvstore& kvstore)
+  : _ntp(ntp)
+  , _name(snapshot_name)
+  , _snapshot_key(ssx::sformat("{}/{}", snapshot_name, ntp))
+  , _log(log)
+  , _kvstore(kvstore) {}
+
+bytes kvstore_backed_stm_snapshot::snapshot_key() const {
+    bytes k;
+    k.append(
+      reinterpret_cast<const uint8_t*>(_snapshot_key.begin()),
+      _snapshot_key.size());
+    return k;
+}
+
+const ss::sstring& kvstore_backed_stm_snapshot::name() { return _name; }
+
+ss::sstring kvstore_backed_stm_snapshot::store_path() const {
+    return _snapshot_key;
+}
+
+ss::future<std::optional<stm_snapshot>>
+kvstore_backed_stm_snapshot::load_snapshot() {
+    auto snapshot_blob = _kvstore.get(
+      storage::kvstore::key_space::stms, snapshot_key());
+    if (!snapshot_blob) {
+        co_return std::nullopt;
+    }
+
+    iobuf_parser snapshot_parser(std::move(*snapshot_blob));
+    const size_t snapshot_size = snapshot_parser.bytes_left();
+    auto header = read_snapshot_header(snapshot_parser, _ntp, name());
+    if (!header) {
+        vlog(_log.warn, "Skipping snapshot {} due to old format", store_path());
+        co_return std::nullopt;
+    }
+
+    stm_snapshot snapshot;
+    snapshot.header = *header;
+
+    // The data portion of the snapshot is all remaining bytes of the blob
+    snapshot.data = snapshot_parser.copy(snapshot_parser.bytes_left());
+    _snapshot_size = snapshot_size;
+    co_return snapshot;
+}
+
+ss::future<>
+kvstore_backed_stm_snapshot::persist_snapshot(stm_snapshot&& snapshot) {
+    iobuf value_buf;
+
+    int8_t version = snapshot_version;
+    int64_t offset = snapshot.header.offset();
+    int8_t data_version = snapshot.header.version;
+    int32_t data_size = snapshot.header.snapshot_size;
+    vassert(
+      size_t(data_size) == snapshot.data.size_bytes(),
+      "Inconsistent snapshot data size");
+    reflection::serialize(value_buf, version, offset, data_version, data_size);
+    value_buf.append(std::move(snapshot.data));
+    const size_t snapshot_size = value_buf.size_bytes();
+
+    co_await _kvstore.put(
+      storage::kvstore::key_space::stms, snapshot_key(), std::move(value_buf));
+    _snapshot_size = snapshot_size;
+}
+
+ss::future<> kvstore_backed_stm_snapshot::remove_persistent_state() {
+    co_await _kvstore.remove(storage::kvstore::key_space::stms, snapshot_key());
+    _snapshot_size = 0;
+}
+
+size_t kvstore_backed_stm_snapshot::get_snapshot_size() const {
+    return _snapshot_size;
+}
```

Comment:
This needs to be verified it may be the case that the size of the snapshot is already accounted in the size of `kvstore` log. In this case this method should return `0`